src/Build/Evaluation/Evaluator.cs

```diff
@@ -1082,21 +1082,57 @@ private void AddBeforeAndAfterTargetMappings(ProjectTargetElement targetElement,
+        /// Ensure the built in property MSBuildDisableChangeWaveVersion is set properly while raising appropriate warnings.
+        /// </summary>
+        /// <returns> String representation of the set change wave, or "999.999" if unset. </returns>
+        private string SanitizeChangeWave()
+        {
+            Version changeWave;
+
+            // If unset, enable all features.
+            if (string.IsNullOrEmpty(Traits.Instance.MSBuildDisableChangeWaveVersion))
+            {
+                Traits.Instance.MSBuildDisableChangeWaveVersion = ChangeWaves.EnableAllFeaturesBehindChangeWaves;
+                return ChangeWaves.EnableAllFeaturesBehindChangeWaves;
+            }
+
+            // If the user-set change wave is of invalid format, log a warning and enable all features.
+            if (!Version.TryParse(Traits.Instance.MSBuildDisableChangeWaveVersion, out changeWave))
+            {
+                _evaluationLoggingContext.LogWarning(null, new BuildEventFileInfo(""), "ChangeWave_InvalidFormat", Traits.Instance.MSBuildDisableChangeWaveVersion);
+                Traits.Instance.MSBuildDisableChangeWaveVersion = ChangeWaves.EnableAllFeaturesBehindChangeWaves;
+                changeWave = ChangeWaves.EnableAllFeaturesVersion;
+            }
+            // If the change wave is out of rotation, log a warning and disable all features.
+            else if (ChangeWaves.IsChangeWaveOutOfRotation(changeWave))
+            {
+                _evaluationLoggingContext.LogWarning(null, new BuildEventFileInfo(""), "ChangeWave_OutOfRotation", ChangeWaves.AllWaves[0], changeWave.ToString());
+                Traits.Instance.MSBuildDisableChangeWaveVersion = ChangeWaves.AllWaves[0];
+                changeWave = ChangeWaves.LowestWaveVersion;
+            }
+
+            return changeWave.ToString();
+        }
+
         /// <summary>
         /// Set the built-in properties, most of which are read-only
         /// </summary>
         private void AddBuiltInProperties()
         {
             string startupDirectory = BuildParameters.StartupDirectory;
 
-            SetBuiltInProperty(ReservedPropertyNames.toolsVersion, _data.Toolset.ToolsVersion);
-            SetBuiltInProperty(ReservedPropertyNames.toolsPath, _data.Toolset.ToolsPath);
-            SetBuiltInProperty(ReservedPropertyNames.binPath, _data.Toolset.ToolsPath);
-            SetBuiltInProperty(ReservedPropertyNames.startupDirectory, startupDirectory);
-            SetBuiltInProperty(ReservedPropertyNames.buildNodeCount, _maxNodeCount.ToString(CultureInfo.CurrentCulture));
-            SetBuiltInProperty(ReservedPropertyNames.programFiles32, FrameworkLocationHelper.programFiles32);
-            SetBuiltInProperty(ReservedPropertyNames.assemblyVersion, Constants.AssemblyVersion);
-            SetBuiltInProperty(ReservedPropertyNames.version, MSBuildAssemblyFileVersion.Instance.MajorMinorBuild);
+            List<P> builtInProperties = new List<P>(20);
```

Comment:
We used to have this list, but one of my optimizations of Evaluator was to stop making (and adding to) a list we never use. I'm guessing @BenVillalobos used an older version of the code when writing this, then resolved the conflicts to take his changes. It would be good to merge them, though.

Reply:
I'll get rid of the list, I didn't realize it wasn't being used.